cms/src/timestamped_data.rs

```diff
@@ -0,0 +1,199 @@
+
+use alloc::{string::String, vec::Vec};
+use const_oid::ObjectIdentifier;
+use der::{
+    asn1::{Ia5String, OctetString, OctetStringRef},
+    Any, Choice, Enumerated, Sequence,
+};
+use spki::AlgorithmIdentifierOwned;
+use x509_cert::{
+    attr::{Attribute, Attributes},
+    crl::CertificateList,
+};
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq, PartialOrd, Ord, Enumerated)]
+#[asn1(type = "INTEGER")]
+#[repr(u8)]
+#[allow(missing_docs)]
+pub enum TsdVersion {
+    V1 = 1,
+}
+
+/// The `TimeStampedData` type is defined in [RFC 5544 Section 2].
+///
+/// ```text
+/// TimeStampedData ::= SEQUENCE {
+///     version              INTEGER { v1(1) },
+///     dataUri              IA5String OPTIONAL,
+///     metaData             MetaData OPTIONAL,
+///     content              OCTET STRING OPTIONAL,
+///     temporalEvidence     Evidence
+/// }
+/// ```
+///
+/// [RFC 5544 Section 2]: https://www.rfc-editor.org/rfc/rfc5544#section-2
+#[derive(Clone, Debug, Eq, PartialEq, Sequence)]
+#[allow(missing_docs)]
+pub struct TimeStampedData<'a> {
+    pub version: TsdVersion,
+    #[asn1(optional = "true")]
+    pub data_uri: Option<Ia5String>,
+    #[asn1(optional = "true")]
+    pub meta_data: Option<MetaData>,
+    #[asn1(optional = "true")]
+    pub content: Option<OctetStringRef<'a>>,
+    pub temporal_evidence: Evidence,
+}
+
+/// ```text
+///  MetaData ::= SEQUENCE {
+///     hashProtected        BOOLEAN,
+///     fileName             UTF8String OPTIONAL,
+///     mediaType            IA5String OPTIONAL,
+///     otherMetaData        Attributes OPTIONAL
+/// }
+/// ```
+#[derive(Clone, Debug, Eq, PartialEq, Sequence)]
+#[allow(missing_docs)]
+pub struct MetaData {
+    pub hash_protected: bool,
+    #[asn1(optional = "true")]
+    pub file_name: Option<String>,
+    #[asn1(optional = "true")]
+    pub media_type: Option<Ia5String>,
+    #[asn1(optional = "true")]
+    pub other_meta_data: Option<Attributes>,
+}
+
+/// ```text
+/// Evidence ::= CHOICE {
+///     tstEvidence    [0] TimeStampTokenEvidence,   -- see RFC 3161
+///     ersEvidence    [1] EvidenceRecord,           -- see RFC 4998
+///     otherEvidence  [2] OtherEvidence
+/// }
+/// ```
+#[derive(Clone, Debug, Eq, PartialEq, Choice)]
+#[allow(missing_docs)]
+pub enum Evidence {
+    #[asn1(context_specific = "0")]
+    TstEvidence(TimeStampTokenEvidence),
+    #[asn1(context_specific = "1")]
+    ErsEvidence(EvidenceRecord),
+    #[asn1(context_specific = "2")]
+    OtherEvidence(OtherEvidence),
+}
+
+/// ```text
+/// TimeStampTokenEvidence ::= SEQUENCE SIZE(1..MAX) OF TimeStampAndCRL
+/// ```
+pub type TimeStampTokenEvidence = Vec<TimeStampAndCRL>;
+
+/// ```text
+/// TimeStampAndCRL ::= SEQUENCE {
+///     timeStamp   TimeStampToken,          -- according to RFC 3161
+///     crl         CertificateList OPTIONAL -- according to RFC 5280
+/// }
+/// ```
+#[derive(Clone, Debug, Eq, PartialEq, Sequence)]
+#[allow(missing_docs)]
+pub struct TimeStampAndCRL {
```

Comment:
```suggestion
pub struct TimeStampAndCrl {
```

https://rust-lang.github.io/api-guidelines/naming.html#casing-conforms-to-rfc-430-c-case